training/src/main/scala/com/airbnb/aerosolve/training/KernelTrainer.scala

```diff
@@ -0,0 +1,126 @@
+package com.airbnb.aerosolve.training
+
+import java.util
+
+import com.airbnb.aerosolve.core.models.KernelModel
+import com.airbnb.aerosolve.core.Example
+import com.airbnb.aerosolve.core.FeatureVector
+import com.airbnb.aerosolve.core.FunctionForm
+import com.airbnb.aerosolve.core.ModelRecord
+import com.airbnb.aerosolve.core.util.StringDictionary
+import com.airbnb.aerosolve.core.util.Util
+import com.typesafe.config.Config
+import org.apache.spark.SparkContext
+import org.apache.spark.rdd.RDD
+import org.slf4j.{Logger, LoggerFactory}
+
+import scala.util.Random
+import scala.util.Try
+import scala.collection.JavaConversions._
+import scala.collection.JavaConverters._
+
+// The decision tree is meant to be a prior for the spline model / linear model
+object KernelTrainer {
+  val log: Logger = LoggerFactory.getLogger("KernelTrainer")
+
+  // Mapping from names to kernel types
+  private final val FunctionFormMap = Map(
+    "rbf" -> FunctionForm.RADIAL_BASIS_FUNCTION,
+    "acos" -> FunctionForm.ARC_COSINE
+  )
+  
+  def train(sc : SparkContext,
+            input : RDD[Example],
+            config : Config,
+            key : String) : KernelModel = {
+    val modelConfig = config.getConfig(key)
+    val candidateSize : Int = modelConfig.getInt("num_candidates")
+
+    val learningRate : Float = Try(modelConfig.getDouble("learning_rate").toFloat).getOrElse(0.1f)
+    val rankKey : String = modelConfig.getString("rank_key")
+    val rankThreshold : Double = Try(modelConfig.getDouble("rank_threshold")).getOrElse(0.0f)
+
+    val examples = LinearRankerUtils
+        .makePointwiseFloat(input, config, key)
+    val model = initModel(modelConfig, examples)
+        
+    val loss : String = modelConfig.getString("loss")
+        
+    val candidates = examples.take(candidateSize)
+
+    // Super simple SGD trainer. Mostly to get the unit test to pass
+    for (candidate <- candidates) {
+      val gradient = computeGradient(model, candidate.example(0), loss, rankKey, rankThreshold)
+      if (gradient != 0.0) {
+        val flatFeatures = Util.flattenFeature(candidate.example(0));
+        model.onlineUpdate(gradient, learningRate, flatFeatures)
+      }
```

Comment:
seems we are fixing all the support vectors the same all the time during the training, perhaps we can delete some support vectors if their responses are very small.